### PR TITLE
zypper: add simple_errors option - fixes #8416

### LINKED ISSUE #8416: [Feature Request][zypper] Add parameter to simplify/reduce error output
### Summary

This is a feature request for the zypper module:

- When updating all system packages using this module and the task fails due to underlying OS issues(disk space).
- Massive XML dump on failure is hard to parse through to pick out relevant error messages
- Request to add option to parse only `<message>` tags in output on failure

### Issue Details

When patching many systems with the zypper module we occasionally hit an issue where installing packages fails due to lack of free space on the client machines.  

There are often many packages to be updated and when these failures occur there is alot of xml output from zypper (and the module) to parse through.  This makes it harder to pick out the actual error message from the output.

Generally on a failure like this, only the text in the `<message>` tags are relevant instead of the entire xml output.

### Proposed Solution

Ideally we could pass a parameter like `simple_errors: true` to the module and it would only dump the text in the `<message>` tags on failure instead of the entire xml output.

I have written some code for this but haven't contributed here before.  Please let me know if this is something that others might be interested in.

### Issue Type

Feature Idea

### Component Name

zypper

### Additional Information

Adding `simple_errors: true` would ensure only text from `<message>` tags are dumped to stout instead of the entire xml output on failure:
```yaml (paste below)
- name: Apply all available patches
  community.general.zypper:
    name: '*'
    state: latest
    type: patch
    simple_errors: true
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by erichoog**
##### SUMMARY
This fixes #8416 

It adds a `simple_errors` option for the zypper module which can be used to simplify or reduce the error output of the zypper module during package installs/updates so that only the content of `<message>` tags are displayed.  (Default is `false` so the default behavior is unaffected)

It also adds a `quiet` option for the zypper module to allow for the option to disable the `--quiet` parameter from being passed to the zypper command. (Default is `true` so the default behavior is unaffected.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
zypper module

### CODE REVIEW

File: plugins/modules/zypper.py [hunk 1]
```diff
@@ -42,7 +42,7 @@
 options:
     name:
         description:
-        - Package name V(name) or package specifier or a list of either.
+       - Package name V(name) or package specifier or a list of either.
```
Comment (felixfontein) [inline plugins/modules/zypper.py:45]:
```suggestion
        - Package name V(name) or package specifier or a list of either.
```

File: plugins/modules/zypper.py [hunk 2]
```diff
@@ -53,7 +53,7 @@
         elements: str
     state:
         description:
-          - V(present) will make sure the package is installed.
+           - V(present) will make sure the package is installed.
```
Comment (felixfontein) [inline plugins/modules/zypper.py:56]:
```suggestion
          - V(present) will make sure the package is installed.
```

File: plugins/modules/zypper.py [hunk 3]
```diff
@@ -144,6 +144,13 @@
         description:
           - Adds C(--clean-deps) option to I(zypper) remove command.
         version_added: '4.6.0'
+    simple_errors:
+        type: bool
+        required: false
+        default: false
+        description:
+          - Provides a simplified error output (parses only the <message> tag text in the XML output)
```
Comment (felixfontein) [inline plugins/modules/zypper.py:152]:
```suggestion
          - When set to V(true), provide a simplified error output (parses only the C(<message>) tag text in the XML output).
```

File: plugins/modules/zypper.py [hunk 4]
```diff
@@ -588,7 +614,7 @@ def main():
         elif state in ['installed', 'present', 'latest']:
             packages_changed, retvals = package_present(module, name, state == 'latest')
 
-    retvals['changed'] = retvals['rc'] in [0, 102] and bool(packages_changed)
+        retvals['changed'] = retvals['rc'] in [0, 102] and bool(packages_changed)
```
Comment (felixfontein) [inline plugins/modules/zypper.py:617]: Is this change intentional?

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)? Thanks.

File: plugins/modules/zypper.py (hunk 4 above)
Comment (erichoog) [reply plugins/modules/zypper.py:617]: This was not intentional, I am not sure why some of these spacing issues were added in as my git diffs don't show that. Will update this accordingly.

Comment (erichoog) [reply plugins/modules/zypper.py]: Submitted another commit fixing this.

**Comment by erichoog** [comment]
> Thanks for your contribution! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)? Thanks.

Changelog fragment added as requested.

File: changelogs/fragments/9270-zypper-add-simple_errors.yaml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - zypper - add simple_errors option (https://github.com/ansible-collections/community.general/pull/9270).
```
Comment (felixfontein) [inline changelogs/fragments/9270-zypper-add-simple_errors.yaml:2]:
```suggestion
  - zypper - add ``simple_errors`` option (https://github.com/ansible-collections/community.general/pull/9270).
```

File: plugins/modules/zypper.py [hunk 6]
```diff
@@ -350,9 +363,21 @@ def parse_zypper_xml(m, cmd, fail_not_found=True, packages=None):
...
+
+    if m.params['simple_errors']:
+        stdout = get_simple_errors(dom)
+
     m.fail_json(msg='Zypper run command failed with return code %s.' % rc, rc=rc, stdout=stdout, stderr=stderr, cmd=cmd)
 
 
+def get_simple_errors(dom):
+    simple_errors = []
+    message_xml_tags = dom.getElementsByTagName('message')
```
Comment (felixfontein) [inline plugins/modules/zypper.py:375]: In case `message_xml_tags` does not exist (it's probably `None` then?) you need to do something other than the `for` loop. How about returning `None`, and changing the invocation so that in case it returns `None`, it doesn't store the result in `stdout`?

File: plugins/modules/zypper.py [hunk 7]
```diff
@@ -350,9 +363,21 @@ def parse_zypper_xml(m, cmd, fail_not_found=True, packages=None):
             return parse_zypper_xml(m, cmd, fail_not_found=fail_not_found, packages=packages)
 
         return packages, rc, stdout, stderr
+
+    if m.params['simple_errors']:
+        stdout = get_simple_errors(dom)
```
Comment (felixfontein) [inline plugins/modules/zypper.py:368]: Changing this to
```suggestion
        stdout = get_simple_errors(dom) or stdout
```
would allow `get_simple_errors()` to return `None` (or an empty string) in case it doesn't find errors, and it will simply use `stdout` in that case.

File: plugins/modules/zypper.py (hunk 6 above)
Comment (erichoog) [reply plugins/modules/zypper.py:377]: Added the change as per my understanding.  Let me know if that works.

**Comment by erichoog** [comment]
Setting this to draft as it requires some additional work and testing.

**Comment by felixfontein** [comment]
Please note that there's now a conflict since we reformatted the module's documentation.

File: plugins/modules/zypper.py [hunk 8]
```diff
@@ -349,15 +370,42 @@ def parse_zypper_xml(m, cmd, fail_not_found=True, packages=None):
             # run zypper again with the same command to complete update
             return parse_zypper_xml(m, cmd, fail_not_found=fail_not_found, packages=packages)
 
+        # apply simple_errors logic to rc 0,102,103,106
+        if m.params['simple_errors']:
+            simple_errors = get_simple_errors(dom)
+            if simple_errors is not None:
+                stdout = simple_errors
```
Comment (felixfontein) [inline plugins/modules/zypper.py:377]: This can be simplified to
```suggestion
            stdout = get_simple_errors(dom) or stdout
```
but if you prefer the longer form (which is easier to understand for folks who aren't that familiar with using `None` in boolean expressions) that's also fine.

**Review by felixfontein — COMMENTED** [review]
Besides this and the conflicts, looks good to me!

**Comment by erichoog** [comment]
I made a mess of this rebase, probably should have been using a feature branch but for some reason I didn't this time.

Any suggestions on best approach to clean this up?

**Comment by felixfontein** [comment]
Can you tick the box that maintainers can push to your branch? (Right now it doesn't seem to be active, at least the tell-tale text is missing.) Then I'll clean it up and force-push it for you.

**Comment by felixfontein** [comment]
If you want to try this yourself, I'd probably try running `git rebase upstream/master` or `git rebase upstream master` (I always forget which of the two is the right syntax) after a `git fetch upstream`, if `upstream` is the remote for ansible-collections/community.general.

**Comment by erichoog** [comment]
> Can you tick the box that maintainers can push to your branch? (Right now it doesn't seem to be active, at least the tell-tale text is missing.) Then I'll clean it up and force-push it for you.

It looks like the box is already ticked:

image

**Comment by erichoog** [comment]
> If you want to try this yourself, I'd probably try running `git rebase upstream/master` or `git rebase upstream master` (I always forget which of the two is the right syntax) after a `git fetch upstream`, if `upstream` is the remote for ansible-collections/community.general.

So I already tried this method but when I do I end up getting a bunch of conflicts which I thought I already resolved.  You are more than welcome to fix it up if you are able to.

File: plugins/modules/zypper.py [hunk 9]
```diff
@@ -586,7 +632,7 @@ def main():
         elif state in ['installed', 'present', 'latest']:
             packages_changed, retvals = package_present(module, name, state == 'latest')
 
-    retvals['changed'] = retvals['rc'] in [0, 102] and bool(packages_changed)
+        retvals['changed'] = retvals['rc'] in [0, 102] and bool(packages_changed)
```
Comment (erichoog) [inline plugins/modules/zypper.py:635]: This was an unintentional change.  Not clear why it is showing indented by may have been caused by my attempt at merge conflict resolution.  I will revert this change.

**Comment by felixfontein** [comment]
> > Can you tick the box that maintainers can push to your branch? (Right now it doesn't seem to be active, at least the tell-tale text is missing.) Then I'll clean it up and force-push it for you.
> 
> It looks like the box is already ticked:
> 
> image

GitHub changed their UI; previously you could detect that the checkbox was set by looking at the bottom of the discussion, GH would mention there that I (as a maintainer) can add additional commits by pushing to repo/branch. That text is no longer there; I just noticed that now there's an entry in the sidebar on the right: "Maintainers are allowed to edit this pull request." That's definitely easier to find, but not if you expect it to be somewhere else ;-)

I'll try to fix this in the next few minutes.

**Comment by felixfontein** [comment]
I think it should be OK now. Can you check? Thanks :)

**Comment by erichoog** [comment]
> I think it should be OK now. Can you check? Thanks :)

Thanks very much!  I'll check if over now!

**Comment by erichoog** [comment]
> I think it should be OK now. Can you check? Thanks :)

Everything is looking good, and my tests looked good as well.  Thanks for all your assistance so far!

**Review by felixfontein — APPROVED** [review]
Looks good from my side (I don't use Zypper, I only looked at the implementation).

**Comment by felixfontein** [comment]
@erichoog thanks for your contribution!
